circuits/Smt.circom

```diff
@@ -1,167 +1,255 @@
+    for (var i = 0; i < depth; i++) {
+        depths.siblings[i] <== siblings[i];
     }
 
-    component sm[nLevels];
-
-    for (i = 0; i < nLevels; i++) {
-        sm[i] = SMTVerifierSM();
+    component nodeType[depth];
+    for (var i = 0; i < depth; i++) {
+        nodeType[i] = NodeTypeDeterminer();
 
         if (i == 0) {
-            sm[i].prev_top <== 1;
+            nodeType[i].previousMiddle <== 1;
+            nodeType[i].previousEmpty <== 0;
+            nodeType[i].previousLeaf <== 0;
+            nodeType[i].previousAuxLeaf <== 0;
+            nodeType[i].previous_na <== 0;
         } else {
-            sm[i].prev_top <== sm[i - 1].st_top;
+            nodeType[i].previousMiddle <== nodeType[i - 1].middle;
+            nodeType[i].previousEmpty <== nodeType[i - 1].empty;
+            nodeType[i].previousLeaf <== nodeType[i - 1].leaf;
+            nodeType[i].previousAuxLeaf <== nodeType[i - 1].auxLeaf;
+            nodeType[i].previous_na <== nodeType[i - 1].current_na;
         }
 
-        sm[i].levIns <== smtLevIns.levIns[i];
+        nodeType[i].auxIsEmpty <== auxIsEmpty;
+        nodeType[i].isExclusion <== isExclusion;
+        nodeType[i].isDesiredDepth <== depths.desiredDepth[i];
     }
 
-    component levels[nLevels];
+    // When isExclusion=0 & auxIsEmpty=1
+    nodeType[depth-1].current_na + nodeType[depth-1].auxLeaf + nodeType[depth-1].leaf + nodeType[depth-1].empty === 1;
 
-    for (i = nLevels - 1; i != -1; i--) {
-        levels[i] = SMTVerifierLevel();
+    component depthHash[depth];
+    for (var i = depth - 1; i >= 0; i--) {
+        depthHash[i] = DepthHash();
 
-        levels[i].st_top <== sm[i].st_top;
-        levels[i].st_inew <== sm[i].st_inew;
+        depthHash[i].isMiddle <== nodeType[i].middle;
+        depthHash[i].isLeaf <== nodeType[i].leaf;
+        depthHash[i].isAuxLeaf <== nodeType[i].auxLeaf;
 
-        levels[i].sibling <== siblings[i];
-        levels[i].new1leaf <== hash1New.out;
+        depthHash[i].sibling <== siblings[i];
+        depthHash[i].auxLeaf <== auxHash.out;
+        depthHash[i].leaf <== hash.out;
 
-        levels[i].lrbit <== n2bNew.out[i];
+        depthHash[i].currentKeyBit <== keyBits.out[i];
 
-        if (i == nLevels - 1) {
-            levels[i].child <== 0;
+        if (i == depth - 1) {
+            depthHash[i].child <== 0;
         } else {
-            levels[i].child <== levels[i + 1].root;
+            depthHash[i].child <== depthHash[i + 1].root;
         }
     }
 
-    levels[0].root === root;
-}
 
-component main {public [root]} = SMTVerifier(10);
+    // Check that if checking for exclusion and auxIsEmpty==0 then key!=auxKey
+    component areKeyEquals = IsEqual();
+    areKeyEquals.in[0] <== auxKey;
+    areKeyEquals.in[1] <== key;
+
+    component keysOk = MultiAND(3);
+    keysOk.in[0] <== isExclusion;
+    keysOk.in[1] <== inverse(auxIsEmpty);
+    keysOk.in[2] <== areKeyEquals.out;
+    keysOk.out === 0;
```

Comment:
```suggestion
    keysOk.in[2] <== inverse(areKeyEquals.out);
    keysOk.out === 1;
```

??

Reply:
In this case, two other values have to be true always.